src/Core/Managed/Shared/ActivityExtensions.cs

```diff
@@ -0,0 +1,55 @@
+﻿#if !NET40
+namespace Microsoft.ApplicationInsights
+{
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Runtime.CompilerServices;
+
+    internal static class ActivityExtensions
+    {
+        private const string OperationNameTag = "OperationName";
+        private static bool isInitialized = false;
+        private static bool isEnabled = false;
+
+        /// <summary>
+        /// Checks that Activity API could be called (DiagnosticSource DLL is loaded).
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoOptimization | MethodImplOptions.NoInlining)] 
+        public static bool IsActivityEnabled()
+        {
+            if (isInitialized)
+            {
+                return isEnabled;
+            }
+
+            try
+            {
+                var activity = new Activity("IsEnabled");
+                isEnabled = true;
+                isInitialized = true;
+                return true;
+            }
+            catch (System.IO.FileNotFoundException)
```

Comment:
It seems a bit indirect to rely on ```FileNotFoundException``` to catch this. I'm curious how you don't end up with a ```TypeLoadException``` when this method is JIT-ted (but the DiagnosticsSource assembly is not present). Is that, perhaps, why you added ```NoOptimization```?

A more direct approach is to try and instantiate the ```Activity``` type either directly via ```Activator.CreateInstance``` using the fully-qualified name or indirectly by calling a method that uses it and catching the ```TypeLoadException```.